teps/0140-producing-results-in-matrix.md

```diff
@@ -0,0 +1,702 @@
+        matrix:
+          params:
+            - name: platform
+              value:
+                - linux
+                - mac
+                - windows
+            - name: browser
+              value:
+                - chrome
+                - safari
+                - firefox
+        taskSpec:
+          params:
+            - name: platform
+              default: ""
+            - name: browser
+              default: ""
+          results:
+            - name: report-url
+              type: string
+          steps:
+            - name: produce-report-url
+              image: alpine
+              script: |
+                echo "Running tests on $(params.platform)-$(params.browser)"
+                echo -n "https://api.example/get-report/$(params.platform)-$(params.browser)" | tee $(results.report-url.path)
+      - name: task-consuming-results
+        taskRef:
+          name: stringtask
+          kind: Task
+        params:
+          - name: url
+            Value: $(tasks.matrix-emitting-results.results.report-url[*])
+        taskSpec:
+          params:
+            - name: url
+              type: string
+          steps:
+            - name: echo
+              image: alpine
+              script: |
+                for arg in "$@"; do
+                  echo "Arg: $arg"
+                done
+```
+
+In the example below, a user wants to run tests on a combination of different platforms and browsers and then fetch the reports for only linux platforms.
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: platforms-with-results
+spec:
+  serviceAccountName: "default"
+  pipelineSpec:
+    tasks:
+      - name: matrix-emitting-results
+        matrix:
+          params:
+            - name: platform
+              value:
+                - linux
+                - mac
+                - windows
+            - name: browser
+              value:
+                - chrome
+                - safari
+                - firefox
+        taskSpec:
+          params:
+            - name: platform
+              default: ""
+            - name: browser
+              default: ""
+          results:
+            - name: report-url
+              type: string
+          steps:
+            - name: produce-report-url
+              image: alpine
+              script: |
+                echo "Running tests on $(params.platform)-$(params.browser)"
+                echo -n "https://api.example/get-report/$(params.platform)-$(params.browser)" | tee $(results.report-url.path)
+      - name: task-consuming-results
+        taskRef:
+          name: stringtask
+          kind: Task
+        params:
+          - name: url
+            value:
+              - $(tasks.matrix-emitting-results.results.report-url[0]) # linux-chrome
+              - $(tasks.matrix-emitting-results.results.report-url[3]) # linux-safari
+              - $(tasks.matrix-emitting-results.results.report-url[6]) # linux-firefox
```

Comment:
What is the use case for dynamically fanning out combinations based on params, but then only using some of those combinations in later tasks? Would matrix include be better suited to these use cases?

Reply:
The syntax here would only work if the param names are static right? If we added a new parameter e.g. `android` to the list then the task would produce unexpected results?

Reply:
> What is the use case for dynamically fanning out combinations based on params, but then only using some of those combinations in later tasks? 

One of them is documented under [Platforms and Browsers](https://github.com/tektoncd/community/blob/4d3cb3b82a9fdb602d91170c752a030faeded9ea/teps/0140-producing-results-in-matrix.md#2-platforms-and-browsers-1) section.

In addition to this, there are many common CI/CD use cases with this kind of requirement in the industry. In microservices world that we are in today, most of the applications comprised of multiple repos, e.g. (1) Source Code (2) Configurations. To bless an application deployment in production, all the repos has to be built using the same `build task` which might be producing a `digest` as a task result. When it comes to deployment task, a digest from building a source repo is needed but not the configuration repo.

There is an additional use case which we discussed this [morning](https://docs.google.com/document/d/16xMubfXC01So-aduTebF8cmyWRcsGYmnHl0EtiIohOY/edit#heading=h.naaqebdi68er) of running the same `scan` task for a given GitHub repo and retrieving only a scan report from the instance which ran scan with `X` to continue execution. The rest of the results might be consumed by the rest of the tasks in a pipeline.

None of these use cases are coming from a single vendor.

> Would matrix include be better suited to these use cases?

Matrix include is designed to specify a combination of `params` for a given `task`. This particular sample pipeline if we re-write using `include`, it will result in a very complex and verbose definition. We will have to spell out all possible combinations of params which were expected to be generated by the controller. And in turn, it defeats the purpose of introducing `matrix`. Include is a very useful feature for the use cases documented in the [TEP](https://github.com/tektoncd/community/blob/main/teps/0118-matrix-with-explicit-combinations-of-parameters.md).

> The syntax here would only work if the param names are static right? 

The syntax here is ideal when the param values are static. Its inherent nature of the `params` which allows `pipeline` and `task` definitions to provide `static` values. We are taking advantage of such static values in our own [CI](https://github.com/tektoncd/pipeline/blob/main/tekton/release-pipeline.yaml).

> If we added a new parameter e.g. `android` to the list then the task would produce unexpected results?
The pipeline author when updates the static value of a `param`, he/she is generally aware of its dependent tasks which are consuming the results. The task would not produce any unexpected results, the pipeline author will have to re-map the indices to match with her/his expectation.

Indexing into an array param/results is not something new being introduced by this TEP. Indexing into an array results is already a beta feature - https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#passing-one-tasks-results-into-the-parameters-or-when-expressions-of-another

All the concerns we have here also applies to the pipeline which consumes array results by indexing into an array or array param from a `pipelineRun` (**without matrix**).

This [example](https://github.com/tektoncd/pipeline/blob/main/examples/v1/pipelineruns/beta/pipelinerun-param-array-indexing.yaml) will result in the same failure or will question the underlying task definition if the order of `environments` changes from `[ 'staging', 'qa', 'prod']` to `[ 'qa', 'prod', 'staging' ]` in `pipelineRun`.

I would like to continue addressing the rest of the comments if we all align on the use case and reach a consensus on this feature request such that @EmmaMunley can focus on updating the TEP if needed.

@lbernick @dibyom @vdemeester @afrittoli thoughts?

Reply:
> One of them is documented under [Platforms and Browsers](https://github.com/tektoncd/community/blob/4d3cb3b82a9fdb602d91170c752a030faeded9ea/teps/0140-producing-results-in-matrix.md#2-platforms-and-browsers-1) section.

To be honest, I'm not sure why someone would want to run tests on different combinations of browsers and platforms, but then only fetch some of the test results. Is there user feedback reflecting this use case?

> To bless an application deployment in production, all the repos has to be built using the same `build task` which might be producing a `digest` as a task result. When it comes to deployment task, a digest from building a source repo is needed but not the configuration repo.

I'm a little bit confused about why you'd want to matrix a single build task for a source repo and a configuration repo. For example, let's say you're using kaniko build for your source repo. Does your configuration repo also contain a Dockerfile that should be built in parallel with the source repo? It seems more likely to me that a configuration repo would contain e.g. infrastructure as code and not a binary that should be built in parallel in the same way as the source repo.

> There is an additional use case which we discussed this [morning](https://docs.google.com/document/d/16xMubfXC01So-aduTebF8cmyWRcsGYmnHl0EtiIohOY/edit#heading=h.naaqebdi68er) of running the same `scan` task for a given GitHub repo and retrieving only a scan report from the instance which ran scan with `X` to continue execution. The rest of the results might be consumed by the rest of the tasks in a pipeline.

Can you elaborate a bit more on why the rest of the pipeline is gated on only one of the scan tasks in this use case? @EmmaMunley it would be helpful to add this use case to the TEP proposal. It may also be helpful to add an example pipeline based on this use case, so we can show what the example would look like with different syntax options.

> > Would matrix include be better suited to these use cases?
> 
> Matrix include is designed to specify a combination of `params` for a given `task`. This particular sample pipeline if we re-write using `include`, it will result in a very complex and verbose definition. We will have to spell out all possible combinations of params which were expected to be generated by the controller.

@EmmaMunley would you be able to write out your example of selecting only certain matrixed results using include instead of indexing so we can compare the syntax in each case?

> Indexing into an array param/results is not something new being introduced by this TEP... This [example](https://github.com/tektoncd/pipeline/blob/main/examples/v1/pipelineruns/beta/pipelinerun-param-array-indexing.yaml) will result in the same failure or will question the underlying task definition if the order of `environments` changes from `[ 'staging', 'qa', 'prod']` to `[ 'qa', 'prod', 'staging' ]` in the `pipelineRun`.

Thanks for this! I thought about it a bit more and this actually makes me more concerned about array indexing rather than making me more convinced we should use it here actually :(
The [TEP](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md) that introduced array indexing explained the use case in this way: "Without indexing, arrays will only work with arrays. This means you cannot combine arrays with Tasks that only take string parameters and this limits the interoperability between Tasks...Use case for array indexing: Unwinding loops (i.e. using one Task multiple times for some set of items)". Now that we can produce matrixed taskruns from array results, I wonder if array indexing is still the best way of doing this?

> I would like to continue addressing the rest of the comments if we all align on the use case and reach a consensus on this feature request such that @EmmaMunley can focus on updating the TEP if needed.

Thanks very much @pritidesai! I would love to see the use cases you've identified here added to the TEP, with any related user feedback if we have it. It sounds like the main goal they have in common is to be able to use a result that was generated with a particular set of parameters; this goal would be helpful to have in the TEP as well, and then we can explore different solutions for achieving this goal.

Reply:
> To be honest, I'm not sure why someone would want to run tests on different combinations of browsers and platforms, but then only fetch some of the test results. Is there user feedback reflecting this use case?

It all depends on how they have configured their pipelines.

Its not about **only fetch some of the test results**. This example could be interpreted as classify and collect the reports for a given `platform` before sending it for further processing. There could be a subsequent set of parallel tasks to collect reports based on the platform and send them to their platform specific buckets.

```
pipelineTask: fetch-and-send-linux-reports
pipelineTask: fetch-and-send-mac-reports
pipelineTask: fetch-and-scan-windows-report
```

As a platform provider, we provide building blocks and its up to the users how they would like to build their choice of resource.